ecs-agent/modeltransformer/transformer.go

```diff
@@ -0,0 +1,106 @@
+// Transform<type> will execute a series of transformation functions to make it compatible with current agent version.
+// add<type>TransformationFunctions will add more <type> transformation functions to the transformation functions chain.
+// Add other transformation functions as needed. e.g. ContainerTransformationFunctions.
+// Add corresponding Transform<Type> and Add<Type>TransformationFunctions while adding other transformation functions.
+type Transformer struct {
+	taskTransformFunctions []*TransformFunc
+}
+
+type transformationFunctionClosure func([]byte) ([]byte, error)
+
+// TransformFunc contains the threshold version string for transformation function and the transformationFunction itself.
+// During upgrade, all models from versions below threshold version should execute the transformation function.
+type TransformFunc struct {
+	version  string
+	function transformationFunctionClosure
+}
+
+func NewTransformer() *Transformer {
+	t := &Transformer{}
+	return t
+}
+
+// GetNumberOfTransformationFunctions returns the number of transformation functions given a model type
+func (t *Transformer) GetNumberOfTransformationFunctions(modelType string) int {
+	switch modelType {
+	case modelTypeTask:
+		return len(t.taskTransformFunctions)
+	default:
+		return 0
+	}
+}
+
+// TransformTask executes the transformation functions when version associated with model in boltdb is below the threshold
+func (t *Transformer) TransformTask(version string, data []byte) ([]byte, error) {
+	var err error
+	// execute transformation functions sequentially and skip those not applicable
+	for _, transformFunc := range t.taskTransformFunctions {
+		if checkVersionSmaller(version, transformFunc.version) {
+			logger.Info(fmt.Sprintf("Agent version associated with task model in boltdb %s is below threshold %s. Transformation needed.", version, transformFunc.version))
+			data, err = transformFunc.function(data)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			logger.Info(fmt.Sprintf("Agent version associated with task model in boltdb %s is bigger or equal to threshold %s. Skipping transformation.", version, transformFunc.version))
+			continue
+		}
+	}
+	return data, err
+}
+
+// AddTaskTransformationFunctions adds the transformationFunction to the handling chain
+func (t *Transformer) AddTaskTransformationFunctions(version string, transformationFunc transformationFunctionClosure) {
+	t.taskTransformFunctions = append(t.taskTransformFunctions, &TransformFunc{
+		version:  version,
+		function: transformationFunc,
+	})
+}
+
+// IsUpgrade checks whether the load of a persisted model to running agent is an upgrade
+func (t *Transformer) IsUpgrade(runningAgentVersion, persistedAgentVersion string) bool {
+	return checkVersionSmaller(persistedAgentVersion, runningAgentVersion)
+}
+
+func checkVersionSmaller(version, threshold string) bool {
+	versionParams := strings.Split(version, ".")
```

Comment:
We can add input validation before line 94, and return false if an input string is invalid.

Reply:
I am not sure if we need an input validation here, since the write is done by agent too (in another save metadata operation)?

Reply:
I think adding a validation is a better practice, and can handle corrupted/modified data.